File: hms-api/src/main/java/com/pms/pharmacy/service/ProductService.java
package com.pms.pharmacy.service;

import com.pms.common.EntityNotFoundException;
import com.pms.pharmacy.dto.ProductDto;
import com.pms.pharmacy.entity.Manufacturer;
import com.pms.pharmacy.entity.Product;
import com.pms.pharmacy.entity.ProductType;
import com.pms.pharmacy.entity.Rack;
import com.pms.pharmacy.repository.ManufacturerRepository;
import com.pms.pharmacy.repository.ProductRepository;
import com.pms.pharmacy.repository.ProductTypeRepository;
import com.pms.pharmacy.repository.RackRepository;
import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository repository;
    private final ProductTypeRepository productTypeRepository;
    private final RackRepository rackRepository;
    private final ManufacturerRepository manufacturerRepository;

    public ProductService(
            ProductRepository repository,
            ProductTypeRepository productTypeRepository,
            RackRepository rackRepository,
            ManufacturerRepository manufacturerRepository) {
        this.repository = repository;
        this.productTypeRepository = productTypeRepository;
        this.rackRepository = rackRepository;
        this.manufacturerRepository = manufacturerRepository;
    }

    public List<ProductDto> findActive() {
        return repository.findByActiveTrueOrderByNameAsc().stream().map(this::toDto).toList();
    }

    public List<ProductDto> findInactive() {
        return repository.findByActiveFalseOrderByUpdatedAtDesc().stream().map(this::toDto).toList();
    }

    @Transactional
    public ProductDto create(ProductDto dto) {
        Product product = new Product();
        applyFields(product, dto);
        product.setActive(true);
        return toDto(repository.save(product));
    }

    @Transactional
    public ProductDto update(Long id, ProductDto dto) {
        Product product = getOrThrow(id);
        applyFields(product, dto);
        return toDto(repository.save(product));
    }

    @Transactional
    public void deactivate(Long id) {
        Product product = getOrThrow(id);
        product.setActive(false);
        repository.save(product);
    }

    @Transactional
    public void restore(Long id) {
        Product product = getOrThrow(id);
        product.setActive(true);
        repository.save(product);
    }

    private void applyFields(Product product, ProductDto dto) {
        ProductType productType = productTypeRepository.findById(dto.productTypeId())
                .orElseThrow(() -> new EntityNotFoundException("Product Type not found: " + dto.productTypeId()));
        Rack rack = rackRepository.findById(dto.rackId())
                .orElseThrow(() -> new EntityNotFoundException("Rack not found: " + dto.rackId()));
        Manufacturer manufacturer = null;
        if (dto.manufacturerId() != null) {
            manufacturer = manufacturerRepository.findById(dto.manufacturerId())
                    .orElseThrow(() -> new EntityNotFoundException("Manufacturer not found: " + dto.manufacturerId()));
        }

        product.setName(dto.name());
        product.setProductType(productType);
        product.setProductCategory(dto.productCategory());
        product.setDrugDosage(dto.drugDosage());
        product.setDrugType(dto.drugType());
        product.setRack(rack);
        product.setManufacturer(manufacturer);
        product.setMedOrNonMed(dto.medOrNonMed());
        product.setCentralGst(dto.centralGst() != null ? dto.centralGst() : 0);
        product.setStateGst(dto.stateGst() != null ? dto.stateGst() : 0);
        product.setHsnSac(dto.hsnSac());
    }

    private Product getOrThrow(Long id) {
        return repository.findById(id).orElseThrow(() -> new EntityNotFoundException("Product not found: " + id));
    }

    private ProductDto toDto(Product product) {
        Manufacturer manufacturer = product.getManufacturer();
        return new ProductDto(
                product.getId(),
                product.getName(),
                product.getProductType().getId(),
                product.getProductType().getName(),
                product.getProductCategory(),
                product.getDrugDosage(),
                product.getDrugType(),
                product.getRack().getId(),
                product.getRack().getName(),
                manufacturer != null ? manufacturer.getId() : null,
                manufacturer != null ? manufacturer.getName() : null,
                product.getMedOrNonMed(),
                product.getCentralGst(),
                product.getStateGst(),
                product.getHsnSac(),
                product.isActive());
    }
}

File: hms-web/src/app/features/pharmacy/inventory-master/products/product.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../../../environments/environment';
import { Product } from './product.model';

export type ProductInput = Pick<
  Product,
  | 'name'
  | 'productTypeId'
  | 'productCategory'
  | 'drugDosage'
  | 'drugType'
  | 'rackId'
  | 'manufacturerId'
  | 'medOrNonMed'
  | 'centralGst'
  | 'stateGst'
  | 'hsnSac'
  | 'scheduleType'
>;

@Injectable({ providedIn: 'root' })
export class ProductService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiBaseUrl}/pharmacy/products`;

  list(): Observable<Product[]> {
    return this.http.get<Product[]>(this.baseUrl);
  }

  listInactive(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.baseUrl}/inactive`);
  }

  create(input: ProductInput): Observable<Product> {
    return this.http.post<Product>(this.baseUrl, input);
  }

  update(id: number, input: ProductInput): Observable<Product> {
    return this.http.put<Product>(`${this.baseUrl}/${id}`, input);
  }

  deactivate(id: number): Observable<void> {
    return this.http.patch<void>(`${this.baseUrl}/${id}/deactivate`, {});
  }

  restore(id: number): Observable<void> {
    return this.http.patch<void>(`${this.baseUrl}/${id}/restore`, {});
  }
}
